src/datachain/hash_utils.py

```diff
@@ -5,99 +5,71 @@
+    if callable(val):
+        return val.__name__ if hasattr(val, "__name__") else str(val)
+    return str(val)
+
+
+def serialize_column_element(expr: Union[str, ColumnElement]) -> dict:
     """
     Recursively serialize a SQLAlchemy ColumnElement into a deterministic structure.
+    Uses SQLAlchemy's _traverse_internals to automatically handle all expression types.
     """
+    from sqlalchemy.sql.elements import BindParameter
 
-    # Binary operations: col > 5, col1 + col2, etc.
-    if isinstance(expr, BinaryExpression):
-        op = (
-            expr.operator.__name__
-            if hasattr(expr.operator, "__name__")
-            else str(expr.operator)
-        )
-        return {
-            "type": "binary",
-            "op": op,
-            "left": serialize_column_element(expr.left),
-            "right": serialize_column_element(expr.right),
-        }
-
-    # Unary operations: -col, NOT col, etc.
-    if isinstance(expr, UnaryExpression):
-        op = (
-            expr.operator.__name__
-            if expr.operator is not None and hasattr(expr.operator, "__name__")
-            else str(expr.operator)
-        )
-
-        return {
-            "type": "unary",
-            "op": op,
-            "element": serialize_column_element(expr.element),  # type: ignore[arg-type]
-        }
-
-    # Function calls: func.lower(col), func.count(col), etc.
-    if isinstance(expr, Function):
-        return {
-            "type": "function",
-            "name": expr.name,
-            "clauses": [serialize_column_element(c) for c in expr.clauses],
-        }
-
-    # Window functions: func.row_number().over(partition_by=..., order_by=...)
-    if isinstance(expr, Over):
-        return {
-            "type": "window",
-            "function": serialize_column_element(expr.element),
-            "partition_by": [
-                serialize_column_element(p) for p in getattr(expr, "partition_by", [])
-            ],
-            "order_by": [
-                serialize_column_element(o) for o in getattr(expr, "order_by", [])
-            ],
-        }
-
-    # Labeled expressions: col.label("alias")
-    if isinstance(expr, Label):
-        return {
-            "type": "label",
-            "name": expr.name,
-            "element": serialize_column_element(expr.element),
-        }
-
-    # Bound values (constants)
+    # Special case: BindParameter has non-deterministic 'key' attribute, only use value
     if isinstance(expr, BindParameter):
         return {"type": "bind", "value": expr.value}
 
-    # Plain columns
-    if hasattr(expr, "name"):
-        return {"type": "column", "name": expr.name}
-
-    # Fallback: stringify unknown nodes
+    # Generic handling for all ColumnElement types using SQLAlchemy's internals
+    if isinstance(expr, ColumnElement):
+        # All standard SQLAlchemy types have _traverse_internals
+        if hasattr(expr, "_traverse_internals"):
```

Comment:
I like that it is somewhat simpler now, but how reliable it is? how maintainable is it ? (they will change the API at some point?) and most likely no one will figure out it and will switch to stringify anyways? So, it is really bad idea to just use plain stringify? I think we discussed this before - it was not obvious, but seemed fine